### PR TITLE
feat: add chunk-dim parameter

### LINKED ISSUE #49: Custom split dimension
### Description

It'd be great to be able to split at dimensions other than 1, enabling the usage of other layers. Currently MemCNN would split the 1st dimension, which is the batch or sequence dimension in case of RNNs, at all times.

### PR DESCRIPTION
**Pull request by ClashLuke**
Closes #49

### CODE REVIEW

**Comment by ClashLuke** [comment]
Please review whether the coding style in the test is alright. I'm not entirely sure if this is what you were looking for. Additionally it may be a good idea to improve the docstring.

File: memcnn/models/tests/test_split_dim.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+from memcnn import AdditiveCoupling, AffineCoupling, InvertibleModuleWrapper
+
+
+class Check(torch.nn.Module):
+    def __init__(self, dim, target_size):
+        super(Check, self).__init__()
+        self.dim = dim
+        self.target_size = target_size
+
+    def forward(self, fn_input: torch.Tensor) -> torch.Tensor:
```
Comment (silvandeleemput) [inline memcnn/models/tests/test_split_dim.py:15]: You can't use typing arguments here since we also support old Python versions (2.7). 
This also shows up in the CI tests.

File: memcnn/models/tests/test_split_dim.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        assert fn_input.size(self.dim) == self.target_size
+        return fn_input
+
+
+@pytest.mark.parametrize('dimension', [None, 1, 2])
+@pytest.mark.parametrize('coupling', [AdditiveCoupling, AffineCoupling])
+def test_split_dim(dimension: int, coupling: Union[AdditiveCoupling, AffineCoupling]):
```
Comment (silvandeleemput) [inline memcnn/models/tests/test_split_dim.py:22]: Similar to above, we can't use typed arguments here.

File: memcnn/models/additive.py [hunk 3]
```diff
@@ -33,7 +33,7 @@ def __init__(self, Fm, Gm=None, implementation_fwd=-1, implementation_bwd=-1, ch
             implementation_bwd : :obj:`int`
                 Switch between different Additive Operation implementations for inverse pass. Default = -1
 
-            chunk_dim : :obj:`int`
+            split_dim : :obj:`int`
                 Dimension used when chunking input tensors. Default = 1
```
Comment (silvandeleemput) [inline memcnn/models/additive.py:37]:
```suggestion
                Dimension to split the input tensors on. Default = 1, generally corresponding to channels.
```

File: memcnn/models/affine.py [hunk 4]
```diff
@@ -81,7 +81,7 @@ def __init__(self, Fm, Gm=None, adapter=None, implementation_fwd=-1, implementat
             implementation_bwd : :obj:`int`
                 Switch between different Affine Operation implementations for inverse pass. Default = -1
 
-            chunk_dim : :obj:`int`
+            split_dim : :obj:`int`
                 Dimension used when chunking input tensors. Default = 1
```
Comment (silvandeleemput) [inline memcnn/models/affine.py:85]:
```suggestion
                Dimension to split the input tensors on. Default = 1, generally corresponding to channels.
```

File: memcnn/models/tests/test_split_dim.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        super(Check, self).__init__()
+        self.dim = dim
+        self.target_size = target_size
+
+    def forward(self, fn_input: torch.Tensor) -> torch.Tensor:
+        assert fn_input.size(self.dim) == self.target_size
+        return fn_input
+
+
+@pytest.mark.parametrize('dimension', [None, 1, 2])
```
Comment (silvandeleemput) [inline memcnn/models/tests/test_split_dim.py:20]: You might also want to test the batch dimension since this should be possible as well.

```suggestion
@pytest.mark.parametrize('dimension', [None, 0, 1, 2])
```

File: memcnn/models/tests/test_split_dim.py (hunk 1 above)
Comment (ClashLuke) [reply memcnn/models/tests/test_split_dim.py:15]: You're right. I'll have to remove them. Unfortunately I cant view the CI information as CircleCI kicks me out of your instance immediately.

File: memcnn/models/tests/test_split_dim.py (hunk 5 above)
Comment (ClashLuke) [reply memcnn/models/tests/test_split_dim.py:20]: I explicitly removed it, as it'd kill the affine coupling. If you can use it with 1 and 2, 0 should work fine (with everything other than affine coupling) as well.

Comment (silvandeleemput) [reply memcnn/models/tests/test_split_dim.py:20]: Interesting, I'll check tomorrow why that's happening.

File: memcnn/models/tests/test_split_dim.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+from typing import Union
```
Comment (silvandeleemput) [inline memcnn/models/tests/test_split_dim.py:1]: Just noticed that this line obviously needs to be removed as well

```suggestion
```

File: memcnn/models/tests/test_split_dim.py (hunk 5 above)
Comment (ClashLuke) [reply memcnn/models/tests/test_split_dim.py:19]: Affine coupling splits the output of the function in two tensors for its multiplication. If we first split it in two and then split it in two different chunks each, it can't work.

Comment (silvandeleemput) [reply memcnn/models/tests/test_split_dim.py:19]: Alright, I see what's going on here. The affine coupling expects the module to output two tensors instead of one. The best solution here is to set the adapter property to the AffineAdapterNaive class when creating the AffineCoupling class. Here are some proposed changes that should make it work (I have tested these changes on my machine):

Add AffineAdapterNaive to the imports from line 5:
```python
from memcnn import AdditiveCoupling, AffineAdapterNaive, AffineCoupling, InvertibleModuleWrapper
```

And modify the test lines 19-end with:
```python
@pytest.mark.parametrize('dimension', [None, 0, 1, 2])
@pytest.mark.parametrize('coupling', [AdditiveCoupling, AffineCoupling])
def test_split_dim(dimension, coupling):
    input_size = (4, 8, 12)
    dim = 1 if dimension is None else dimension
    module = Check(dim, input_size[dim] // 2)
    coupling_args = dict(adapter=AffineAdapterNaive) if coupling.__name__ == 'AffineCoupling' else dict()
    if dimension is not None:
        coupling_args["split_dim"] = dimension
    model = coupling(module, **coupling_args)
    model = InvertibleModuleWrapper(model)
    inp = torch.randn(input_size, requires_grad=False)
    output = model(inp)
    assert inp.shape == output.shape

```

**Comment by silvandeleemput** [comment]
@ClashLuke I have had a look at the pytest this morning and I have posted a solution to the issue in the comment above.

File: memcnn/models/tests/test_split_dim.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+@pytest.mark.parametrize('dimension', [None, 1, 2])
+@pytest.mark.parametrize('coupling', [AdditiveCoupling, AffineCoupling])
+def test_split_dim(dimension, coupling):
+    dim = 1 if dimension is None else dimension
+    module = Check(dim, 2)
+    model = (coupling(module) if dimension is None
+             else coupling(module, split_dim=dimension))
+    model = InvertibleModuleWrapper(model)
+    inp = torch.randn((2, 4, 4), requires_grad=False)
+    model(inp)
```
Comment (silvandeleemput) [inline memcnn/models/tests/test_split_dim.py:28]:
```suggestion
    input_size = (4, 8, 12)
    dim = 1 if dimension is None else dimension
    module = Check(dim, input_size[dim] // 2)
    coupling_args = dict(adapter=AffineAdapterNaive) if coupling.__name__ == 'AffineCoupling' else dict()
    if dimension is not None:
        coupling_args["split_dim"] = dimension
    model = coupling(module, **coupling_args)
    model = InvertibleModuleWrapper(model)
    inp = torch.randn(input_size, requires_grad=False)
    output = model(inp)
    assert inp.shape == output.shape
```

File: memcnn/models/tests/test_split_dim.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
+
+import pytest
+import torch
+
+from memcnn import AdditiveCoupling, AffineCoupling, InvertibleModuleWrapper
```
Comment (silvandeleemput) [inline memcnn/models/tests/test_split_dim.py:5]:
```suggestion
from memcnn import AdditiveCoupling, AffineAdapterNaive, AffineCoupling, InvertibleModuleWrapper
```

**Comment by ClashLuke** [comment]
Now that splitting at the batch dimension is possible, should we add different input sizes as a test as another test, just to be sure?

File: memcnn/models/tests/test_split_dim.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+@pytest.mark.parametrize('coupling', [AdditiveCoupling, AffineCoupling])
+def test_split_dim(dimension, coupling):
+    input_size = (4, 8, 12)
+    dim = 1 if dimension is None else dimension
+    module = Check(dim, input_size[dim] // 2)
+    coupling_args = dict(adapter=AffineAdapterNaive) if coupling.__name__ == 'AffineCoupling' else dict()
+    if dimension is not None:
+        coupling_args["split_dim"] = dimension
+    model = coupling(module, **coupling_args)
+    model = InvertibleModuleWrapper(model)
```
Comment (silvandeleemput) [inline memcnn/models/tests/test_split_dim.py:29]: The CI is failing on the last assert because the InvertibleModuleWrapper is being used and it puts the inp.shape to [0] in PyTorch version 0.4. It is better to remove it since it is not essential to the part being tested and will avoid this error.

File: memcnn/models/tests/test_split_dim.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
+
+import pytest
+import torch
+
+from memcnn import AdditiveCoupling, AffineAdapterNaive, AffineCoupling, InvertibleModuleWrapper
```
Comment (silvandeleemput) [inline memcnn/models/tests/test_split_dim.py:5]: The InvertibleModuleWrapper is not needed for this test and removing it will avoid some CI tests failures. See below for more detailed rationale.
```suggestion
from memcnn import AdditiveCoupling, AffineAdapterNaive, AffineCoupling
```

File: memcnn/models/tests/test_split_dim.py (hunk 9 above)
Comment (ClashLuke) [reply memcnn/models/tests/test_split_dim.py:29]: You're right, the invertible module wrapper isn't necessary.

**Comment by silvandeleemput** [comment]
I am now in the process of releasing version 1.4.0 of MemCNN which includes this PR. Many thanks for  your help. I have added a mention in the HISTORY.rst.
